### PR TITLE
nixos/kibana: add passwordFile option (#24288)

### PR DESCRIPTION
**Pull request by schneefux**
###### Motivation for this change
Added a warning if the elasticsearch password is stored in the nix store, added the `passwordFile` option to be used instead.
Tested Kibana 5 by looking at the generated `/var/lib/kibana/config.json`, once with the `password` option set and once using the `passwordFile` option.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/search/kibana.nix [hunk 1]
```diff
@@ -182,10 +188,22 @@ in {
...
+      preStart = let
+        cfgFileRun = "${cfg.dataDir}/config.json";
+      in ''
+        cp -f ${cfgFile} ${cfgFileRun}
+        ${optionalString (cfg.elasticsearch.passwordFile == null) ''
+          sed -e "s,\"PASSWORD_REPLACED_PRESTART\",null,g" -i ${cfgFileRun}
+        ''}
+        ${optionalString (cfg.elasticsearch.passwordFile != null) ''
+          ELASTIC_PASSWORD=$(head -n1 ${cfg.elasticsearch.passwordFile})
+          sed -e "s,PASSWORD_REPLACED_PRESTART,$ELASTIC_PASSWORD,g" -i ${cfgFileRun}
```
Comment (schneefux) [inline nixos/modules/services/search/kibana.nix]: This will produce invalid JSON if there is a `"` in the password. Is there a better solution than `sed`?

**Comment by schneefux** [comment]
Tested & fixed the expression producing an error when no password was set.

File: nixos/modules/services/search/kibana.nix [hunk 2]
```diff
@@ -182,10 +188,22 @@ in {
...
+      preStart = let
+        cfgFileRun = "${cfg.dataDir}/config.json";
+      in ''
+        cp -f ${cfgFile} ${cfgFileRun}
+        ${optionalString (cfg.elasticsearch.passwordFile == null) ''
+          sed -e 's,"password":"PASSWORD_REPLACED_PRESTART"\,,,g' -i ${cfgFileRun}
+        ''}
+        ${optionalString (cfg.elasticsearch.passwordFile != null) ''
+          ELASTIC_PASSWORD=$(head -n1 ${cfg.elasticsearch.passwordFile})
+          sed -e "s,PASSWORD_REPLACED_PRESTART,$ELASTIC_PASSWORD,g" -i ${cfgFileRun}
```
Comment (schneefux) [inline nixos/modules/services/search/kibana.nix]: This fails if there are special characters like `"` in the password. Is there a better approach?

File: nixos/modules/services/search/kibana.nix [hunk 3]
```diff
@@ -182,10 +188,22 @@ in {
...
       description = "Kibana Service";
       wantedBy = [ "multi-user.target" ];
+      preStart = let
+        cfgFileRun = "${cfg.dataDir}/config.json";
+      in ''
+        cp -f ${cfgFile} ${cfgFileRun}
+        ${optionalString (cfg.elasticsearch.passwordFile == null) ''
+          sed -e 's,"password":"PASSWORD_REPLACED_PRESTART"\,,,g' -i ${cfgFileRun}
+        ''}
+        ${optionalString (cfg.elasticsearch.passwordFile != null) ''
```
Comment (Infinisil) [inline nixos/modules/services/search/kibana.nix]: Use

```nix
if (cfg.elasticsearch.passwordFile == null) then ''
  sed ...
'' else ''
  ELASTIC_PASSWORD=...
''
```

instead

File: nixos/modules/services/search/kibana.nix (hunk 2 above)
Comment (Infinisil) [reply nixos/modules/services/search/kibana.nix]: I don't think this will fail, bash doesn't interpret the quotes in an expanded string.

**Review by infinisil — CHANGES_REQUESTED** [review]

**Review by matthewbauer — CHANGES_REQUESTED** [review]
See @Infinisil's comments

**Comment by schneefux** [comment]
Sorry, I am no longer using Kibana on NixOS so I cannot maintain this pull request. If anyone wants to take over, please do!

**Pull request closed without merging**